cos_registration_server/api/views.py

```diff
@@ -259,3 +266,119 @@ def delete(self, request: Request, uid: str) -> Response:
+    """PrometheusAlertRuleFiles API view."""
+
+    def get(self, request: Request) -> Response:
+        """Prometheus Alert Rules get view.
+
+        request: Http GET request.
+        return: Http JSON response.
+        """
+        # retrieve alert rules that are not a template and serialize them
+        no_template_alert_rules = PrometheusAlertRuleFile.objects.filter(
+            template=False
+        )
+
+        serialized = PrometheusAlertRuleFileSerializer(
+            no_template_alert_rules, many=True
+        )
+
+        # retrieve template alert rules and render them
+        rendered_rules = []
+
+        template_alert_rules = PrometheusAlertRuleFile.objects.filter(
+            template=True
+        )
+        devices = Device.objects.all()
+
+        for device in devices:
+            for rule in device.prometheus_alert_rule_files.all():
+                if rule in template_alert_rules:
+                    env = Environment(
+                        variable_start_string="%%",
+                        variable_end_string="%%",
+                    )
+                    rule_string = yaml.dump(
+                        rule.rules, default_flow_style=False
+                    )
+                    template = env.from_string(rule_string)
+                    context = {"juju_device_uuid": f"{device.uid}"}
+                    rendered_rule = template.render(context)
+                    rendered_rules.append(
+                        {
+                            "uid": rule.uid + "/" + device.uid,
+                            "rules": rendered_rule,
+                        }
+                    )
+
+        # rendered rules are already dumped to get them rendered via jinja
+        # hence they are already serialized as strings.
+        serialized_list = list(serialized.data) + rendered_rules
+        return Response(serialized_list)
+
+    def post(self, request: Request) -> Response:
+        """Prometheus Alert Rules post view.
+
+        request: Http POST request.
+        return: Http JSON response.
+        """
+        serialized = PrometheusAlertRuleFileSerializer(data=request.data)
+        if serialized.is_valid():
+            serialized.save()
+            return Response(serialized.data, status=201)
+        return Response(serialized.errors, status=400)
+
+
+class PrometheusAlertRuleFileView(APIView):
+    """PrometheusAlertRuleFile API view."""
+
+    def _get_alert_rule(self, uid: str) -> PrometheusAlertRuleFile:
+        try:
+            alert_rule = PrometheusAlertRuleFile.objects.get(uid=uid)
+            return alert_rule
+        except PrometheusAlertRuleFile.DoesNotExist:
+            raise NotFound("Object does not exist")
+
+    def get(self, request: Request, uid: str) -> HttpResponse:
+        """Prometheus alert rule get view.
+
+        request: Http GET request.
+        return: Http JSON response.
+        """
+        alert_rule = self._get_alert_rule(uid)
```

Comment:
Shouldn't the potential exception be caught here?

Reply:
There is no need, in django raising NotFound results in a 404 Not Found response. Right now, if we you try to get a non existent UID rule, it will return a 404 response with "object does not exist"